openssh: 7.9p1 -> 8.1p1

dtzWill opened the pull request:
###### Motivation for this change

https://www.openwall.com/lists/oss-security/2019/04/18/1

Sending to staging for anticipated build impact,
please promote to master as needed for security implications.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

adisbladis commented:
I think this warrants a release note.

FRidh commented:
I agree with @adisbladis . Could you add it @dtzWill ?

FRidh reviewed (changes requested):
Release note is needed.

dtzWill committed:
nixos: add release note for openssh upgrade \o/

c0bw3b commented on nixos/doc/manual/release-notes/rl-1909.xml:
```diff
@@ -135,6 +135,14 @@
+    <para>
+     OpenSSH has been upgraded from 7.9 to 8.0, improving security and adding features
+     but with potential incompatibilities.  Consult the
+     <link xlink:href="https://www.openwall.com/lists/oss-security/2019/04/18/1">release announcement</link>
```
Or directly from https://www.openssh.com/txt/release-8.0
(nitpick)

dtzWill commented:
Well the good news is now I kinda maybe know enough doxygen to do this more quickly in the future ;).

Oh, blargh, merge conflict? Gimme a few...

aneeshusa commented:
Thanks for this! I will try to test this in the next few days.
Can you also bump the debian GSSAPI patch? I don't use it myself but I know there are some folks who do. https://salsa.debian.org/ssh-team/openssh/raw/d7e7059a085282da470bd0a4b40b60f01c16806d/debian/patches/gssapi.patch should work.

edef1c committed:
openssh: don't let configure override SSH_KEYSIGN

While 9fe10288f01984963faf47e21bf1bae4d7d37962 ensured that the
ssh-keysign path is searched for in PATH if not absolute,
it doesn't prevent the configure script from defaulting to an
absolute path in $out/libexec, making the whole effort rather
pointless.

dtzWill committed:
openssh: 7.9p1 -> 8.1p1

https://www.openwall.com/lists/oss-security/2019/04/18/1

edef1c commented:
My proposal for `hpnSupport`, after reading logs of [IRC discussion in #nixos-security](https://logs.nix.samueldr.com/nixos-security/2019-08-08#1565278628-1565293359;hpn) and talking to @flokli: We set `meta.broken = hpnSupport` for now, and watch who screams about it breaking. This is easy to revert, and gives a clear signal about our willingness to support the (unmaintained) HPN patches. If that goes well, before releasing 19.10, we remove the `hpnSupport` argument. The latter should go on the list of tasks to cover before we cut the release.

flokli commented:
@edef1c can you push a new version of this PR implementing the suggested changes, and rebase on latest master?

edef1c committed:
openssh: mark hpnSupport as broken

We're hoping to deprecate HPN support, given that as far as we
can tell, nobody is using it, and the patches seem rather unmaintained.

edef1c commented:
> @edef1c can you push a new version of this PR implementing the suggested changes, and rebase on latest master?

That turns out to be a poor idea (sorry for the mess), but rebase onto the latest staging has happened now.

flokli commented:
@edef1c can you move the release notes to 20.03? LGTM otherwise.

dtzWill commented:
Thank you for your continued work on this @edef1c .  For those following along at home, we should now move to 8.1 (having missed 8.0 while it at current :cry:): https://www.openssh.com/txt/release-8.1

Easy update from this, for example: c25290098f83bf422b2856027a26ce2aea620628 .  Especially if we're not anchored by the hpn version (while having no one interested in maintaining it) ;).

flokli commented:
Let's cherry-pick c25290098f83bf422b2856027a26ce2aea620628 on top of that and merge it in.

edef1c commented:
Do we want to do that off this PR? It'll be a little weird to merge an openssh-8p1 branch (which GitHub puts in the merge commit message) and upgrade us to OpenSSH 8.1p1.

flokli commented:
@edef1c this happens with other longer-running PRs aswell.

edef1c commented:
I'm merging this alongside a small fix for a deficiency in 9fe10288f01984963faf47e21bf1bae4d7d37962, since I'd rather not cause two rebuilds of that magnitude.
